ERCS/erc-7821.md

```diff
@@ -0,0 +1,204 @@
+
+Most layer 2s already include calldata compression and forward the calldata savings to users.
+
+As such, we opt for simplicity to enable maximum compatibility and ease of implementation.
+
+Using a standard struct also helps with auditability, block explorer integrations, browser wallet integrations, and [EIP-712](./eip-712) signature generations.
+
+### Usage of a single `opData` bytes argument
+
+The single `opData` bytes argument also provides all the flexibility that is needed.
+
+## Backwards Compatibility
+
+No backwards compatibility issues.
+
+## Reference Implementation
+
+```solidity
+// SPDX-License-Identifier: CC0-1.0
+pragma solidity ^0.8.4;
+
+/// @notice Minimal batch executor mixin.
+abstract contract MinimalBatchExecutor {
+    /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+    /*                          STRUCTS                           */
+    /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+    /// @dev Call struct for the `execute` function.
+    struct Call {
+        address target;
+        uint256 value;
+        bytes data;
+    }
+
+    /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+    /*                   FUNCTIONS TO OVERRIDE                    */
+    /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+    /// @dev Ensures that `execute` can only be called by the correct caller or `opData`.
+    function _authorizeExecute(Call[] calldata calls, bytes calldata opData) internal virtual;
+
+    /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+    /*                    EXECUTION OPERATIONS                    */
+    /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+    /// @dev Executes the `calls` and returns the results.
+    /// Reverts and bubbles up error if any call fails.
+    function execute(Call[] calldata calls, bytes calldata opData)
+        public
+        payable
+        virtual
+        returns (bytes[] memory results)
+    {
+        _authorizeExecute(calls, opData);
+        return _execute(calls);
+    }
+
+    /// @dev This function is provided for frontends to detect support.
+    function minimalBatchExecutorVersion() public pure virtual returns (uint256) {
+        return 1; // This number may change in the future.
+    }
+
+    /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+    /*                      INTERNAL HELPERS                      */
+    /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+    /// @dev Executes the `calls` and returns the results.
+    /// Reverts and bubbles up error if any call fails.
+    function _execute(Call[] calldata calls) internal virtual returns (bytes[] memory results) {
+        results = new bytes[](calls.length);
+        for (uint256 i; i < calls.length; ++i) {
+            Call calldata c = calls[i];
+            results[i] = _execute(c.target, c.value, c.data);
```

Comment:
Suggestion: consider adding a special address (e.g., `address(0)`) where the contract would execute the transaction to self. This is a neat optimization when you must batch multiple calls on the account contract itself.

This is implemented in Safe's MultiSend contract: https://github.com/safe-global/safe-smart-account/blob/7f79aaf05c33df71d9cb687f0bc8a73fa39d25d5/contracts/libraries/MultiSend.sol#L50C17-L50C57

Reply:
Ok, can add. Good for calldata compression.